### DISCUSSION TITLE
JoltGeneric6DOFJoint3D sometimes has an offset on unfreeze

### DISCUSSION THREAD

**Discussion by Conz3D** [godot-jolt/godot-jolt#654] [General]
Before opening an Issue, I want to pin my problem further down.
I have an object (scene) with two RigidBodies connected by a JoltGeneric6DOFJoint3D as a fixed joint. I use the Jolt version, bec. I want to disable the joint at runtime. I spawn the object as frozen and unfreeze on click. _"Exclude Nodes from Collision"_ is activated, so there is no collision between the two bodies.
When placed in the scene in the editor it works as expected 100% of the time, but if I instantiate this object ~1 in 10 times the two rigid bodies have an offset in the joint on unfreezing them. It's visible in the behavior, so it's not only a visual glitch. 

Debugging physics is really problematic. It's hard to tell if it's a godot problem with the processing order, a problem in the godot-jolt binding or in the jolt physic simulation itself.
Running it with the godot physics it's working 100% of the time. (Obviously without the disabling of the joint, since it's only available in Jolt). **So it must be a problem with Jolt or the godot-jolt binding somewhere.**

To be more specific, The object/scene is a rocket with a tip attached by a JoltGeneric6DOFJoint3D (In my case the "TipJoint", set as fixed by not setting the limits or springs).

![scene_rocket](https://github.com/godot-jolt/godot-jolt/assets/129608680/acfa6098-a6d0-4cfc-adcd-cb503642f8e4)

Some rockets are placed in the editor in the scene, and there is a "Spawner", that will respawn a rocket some seconds after it's spawned rocked got started. The placed and spawned rockets are in freeze mode, to not fall over on spawning. They are not placed insinde the floor or the spawner.

![rocket_setup](https://github.com/godot-jolt/godot-jolt/assets/129608680/fc056edb-0e68-4691-a05e-a71948415be7)

In the runtime, by clicking on the rocket it will start.
- Unfreeze the rocket and the tip
- applying a force in the oposite direction of the thruster (rocket motor)

After some time (timer) the thrust will be stopped
- Stop the force and thruster particle
- **Disable the joint**
- play an explosion sound and particle

https://github.com/godot-jolt/godot-jolt/assets/129608680/1475330f-28ba-46b7-8a96-e2c9d5a1d14b

As you can see, all the placed rockets in the editor are working as expected, all the time.
The spawned ones are working most of the time, but every ~10th rocket is bugged. In the video all are fine, but the last one has the bug. The tip is ~20cm in front of the rocket and it's tumbling as if the tip is offset in the physics, too.

![rocket_bug](https://github.com/godot-jolt/godot-jolt/assets/129608680/104274c7-fb8a-45ae-8f9d-7a08e04f6c1b)

Any idea what I could test or could do to fix this?

**Comment by mihe** [comment]
I don't see anything in your setup that immediately strikes me as problematic, although I am a little bit suspicious of the fact that `RigidBody3DTip` is childed to `Rocket`, but the problem would presumably be more consistent if that was the cause.

I'd prefer if you uploaded your project (without all the irrelevant stuff) so that I can reproduce it on my end, otherwise this will likely be quite painful to debug.

**Comment by Conz3D** [comment]
> I don't see anything in your setup that immediately strikes me as problematic, although I am a little bit suspicious of the fact that `RigidBody3DTip` is childed to `Rocket`, but the problem would presumably be more consistent if that was the cause.
> 

IMHO it is more explanatory with the joined body as a child of the joint. I know it's not necessary, but found in my tests not be problem with this hierarchy. But I'll try a version where the 2nd body is a direct child of the root node.

> I'd prefer if you uploaded your project (without all the irrelevant stuff) so that I can reproduce it on my end, otherwise this will likely be quite painful to debug.

Sure, I will do this a little bit later today.

**Comment by Conz3D** [comment]
> I don't see anything in your setup that immediately strikes me as problematic, although I am a little bit suspicious of the fact that `RigidBody3DTip` is childed to `Rocket`, but the problem would presumably be more consistent if that was the cause.

What would be the "perfect" hierachy in this case? I prefere the main RigidBody as the root node of the object/scene.
Would it be better to have a Node3D as root node with two rigid Bodies and the joint as direct childs? Or can be the 2nd Body and the joint both be a direct child of the first RigidBody? Or is this "sus", too?

**Reply by mihe** [reply]
> What would be the "perfect" hierachy in this case?

I'm not sure there's a perfect hierarchy necessarily. The one you have looks fine. I'm by no means an expert on the scene tree, but explicitly modifying the transform of a parent/ancestor to `RigidBody3D` typically means that the `RigidBody3D` will move as well, which can be a surprise to some. Where you put the joint node in the tree shouldn't matter at all. The joint isn't really a thing that persists in the 3D world per se. The only thing that matters is where the joint is in relation to its connected bodies when it's first added to the tree.

**Comment by Conz3D** [comment]
> I'd prefer if you uploaded your project (without all the irrelevant stuff) so that I can reproduce it on my end, otherwise this will likely be quite painful to debug.

@mihe, **here is the scene with only the rockets.**

[godot_jolt_joint_bug.zip](https://github.com/godot-jolt/godot-jolt/files/13061073/godot_jolt_joint_bug.zip)

 Click on the rockets (the yellow part) to start them. There are the in engine placed ones, and the "Rocket spawner". The bug happens only on the rockets of the spawner.
You have to start rockets until you get one with the bug. Sometimes I have to launch 20-30, sometimes it's the 2nd or 3rd one.

_For future uploads, do I have to include the addon (ogdot-jolt) in the project? Without  there would be an error. When jolt is included, you only need to restart godot on first start._

**Reply by mihe** [reply]
> For future uploads, do I have to include the addon (ogdot-jolt) in the project?

No. In fact, I would prefer if you (in the future) trim the repro even more. There's no need for sounds and particle effects and all that stuff. It just adds to the "noise" and makes the issue harder to debug. Even the clicking on the rockets could be replaced with a timer or something.

**Comment by Conz3D** [comment]
> I don't see anything in your setup that immediately strikes me as problematic, although I am a little bit suspicious of the fact that `RigidBody3DTip` is childed to `Rocket`, but the problem would presumably be more consistent if that was the cause.

I have tested a different hierachy. With the main body as root, and the joint and the 2nd RigidBody as direct childs of it. I get the same effect/bug.

![scene_rocket02](https://github.com/godot-jolt/godot-jolt/assets/129608680/dbf2e05f-23bd-4b17-bdc8-3413e200e7b3)

Another hierachy is all as a direct children of a Node3D. There the bug seems not to happen. But it has a very bad and unexpexted flight behavior.
 ![scene_rocket03](https://github.com/godot-jolt/godot-jolt/assets/129608680/99e380e2-5c3a-43d8-8b14-f53a602978cc)

And another one, with the joint as a child of the base RigidBody:
![scene_rocket04](https://github.com/godot-jolt/godot-jolt/assets/129608680/f506ed69-aa56-4fc5-bc2f-fe7293bfea3c)
But it is the same as the one before. No bug, but bad flight behavior.

Have to play with the last ones a little bit more, don't understand the bad flight behavior. 
But it seems you are on the right track, that there is a problem with the hierachy. Something with the order of physics calculation on the RigidBodies.
Will investigate further and report back.

**Reply by mihe** [reply]
I'm having trouble reproducing what you're seeing. I trimmed it down a bit and made it continuously spam the rockets, and I'm still not seeing it.

Are you able to reproduce the issue in this trimmed down version of the project?: [timed_rockets.zip](https://github.com/godot-jolt/godot-jolt/files/13061231/timed_rockets.zip)

I changed some timing values here and there, and put the rockets on their own layer so they don't collide with eachother. It should be pretty much the same besides that.

**Reply by Conz3D** [reply]
Indeed, in your reduced version I can't see the bug. That looks fine. Maybe the timer spawns at a different process step than the input event, I have to investigate the differences further.

Thanks so far.

**Comment by Conz3D** [comment]
> I changed some timing values here and there, and put the rockets on their own layer so they don't collide with eachother. It should be pretty much the same besides that.

**Ok, that seems to fix my issue.** So it looks like it's self colliding of the two rigid bodies, before the joint is active where the collision of the joined bodies is deactivated. 
Good to know, that you can't spawn objects with a joint where the rigid bodies can collide, ignoring the settings in the joint for one frame (or a physiks tick), so the joined body get's an impulse once, but than it's joined.

I think, that's why the bug didn't appear in the rockets that are placed in the editor, there the joint is active at the beginning. But in the spawned one the collision can happen for a shortest amount of time.

Do you think this is a bug to add to the issue tracker? IMHO it's more like a glitch than a bug, but it can lead to a bug in a scene.

Thank you for your help.

EDIT: Problem isn't fixed!

**Reply by mihe** [reply]
> Good to know, that you can't spawn objects with a joint where the rigid bodies can collide, ignoring the settings in the joint for one frame (or a physiks tick), so the joined body get's an impulse once, but than it's joined.

I'll need to take a closer look at the flow of things in your repro, but this shouldn't be possible. The collision exclusions are added on the joint node's `_ready` call, so there shouldn't be a physics tick where the bodies aren't excluded.

> I think, that's why the bug didn't appear in the rockets, that are placed in the editor, there the joint is active at the beginning. But in the spawned one the collision can happen for a shortest amount of time.

The joints don't do anything in the editor. They're technically added to the simulation, but since the simulation in the editor is never stepped, they are largely a no-op there.

> Do you think this is a bug to add to the issue tracker? IMHO it's more like a glitch than a bug, but it can lead to a bug in a scene.

I'll add it if needed.

**Comment by Conz3D** [comment]
Sorry to say, that this problem is not fixed. My first impression was wrong. :(
At least I can say that it has nothing to do with freeze/unfreeze. It still happening when I never set them to frozen.
Still it never happened to the rockets placed in the editor by hand, only with instantiated rockets.
I have tried to set different priorities and physic priorities for the different elements, with no effect.
It seems like the joint is not activated for some frames at the beginning and gets activated with some delay, but the "Enabled" is set to true  in the rocket scene.
Will try again with different hierachies. 
So strange that it's not happening every time, only now and then.

**Comment by Conz3D** [comment]
I'm now counting the correct and the bugged instances.
The more the physics has to do, the more often the bug appears. The physics seems to struggle with setting up the simulation of the joints when calculating too much other joints. Could this be possible?
In a scene where the rockets will come to a sleeping state very quick, I have only 1 in 100 bugged instances. In the same scene with more active rockets at the same time I can bring it up to every 2nd rocket is bugged. And it seems that the distribution of correct and bugged rockets comes in waves. I get a lot of correct starts, and than some bugged in a row, than correct ones again aso. 
There seems to be a clear correlation betw. active (not sleeping) physics objects and the appearence of my problem.

**Comment by Conz3D** [comment]
FYI:
I have started a discussion about this on the jolt git:
https://github.com/jrouwe/JoltPhysics/discussions/769

**Comment by mihe** [comment]
I managed to reproduce it this time for some reason, so I reduced it down to something more manageable once again: [nine_rockets.zip](https://github.com/godot-jolt/godot-jolt/files/13432364/nine_rockets.zip)

More importantly I got rid of the `randf` stuff and moved all the timers to the physics tick rather than the idle tick, to help produce a more deterministic output.

This version will fire off 9 rockets at 2 second intervals, without needing to provide any input, and on the 9th rocket you should hopefully always see the bug happen, assuming you run it on Windows and Godot 4.1.3, but pretty much any change to the initial conditions will seemingly ruin the ability to reproduce it.

While this reduced repro isn't the best place to showcase this, it seems like `ObjectRemover` has some role to play in this whole thing, as this problem seems to go away completely if its `queue_free` never happens.

I don't have time right now to do any actual debugging of what's causing this on the code side of things, but at least now I have a solid repro.

**Reply by Conz3D** [reply]
Thank you for taking your time to look into it again. I'm happy that you can reproduce it now.
I have tested your nine_rockets on Windows with Godot 4.1.3 and exactly the 9th one bugged out.
Take your time to investigate it further whenever it fits. I'm relieved, that you confirmed this as a problem/bug for now.

I have no clue how it can be related to queue_free and the exact setup of my scene.
If I can do anything to support your debugging, post it here. Also I'll have an eye on the issue tracker if you decide that this is an issue to track, when you have a grasp what's happening exactly.

**Reply by mihe** [reply]
I messed around with it some more, and it seems like this isn't limited to 6DOF joints. I can reproduce this just as well when using a `JoltHingeJoint3D` with a zero limit span, which will then internally build Jolt's `FixedConstraint`, which shouldn't be able to have any degrees-of-freedom at all.

I went ahead and added some picking as well, and also disabled the thrust and gravity for the offending (9th) rocket, so that I can actually mess around with it, which you'll find here: [nine_pickable_rockets.zip](https://github.com/godot-jolt/godot-jolt/files/13443391/nine_pickable_rockets.zip)

This is what it looks like when throwing around that rocket, which again is using Jolt's `FixedConstraint`:

https://github.com/godot-jolt/godot-jolt/assets/4884246/902095fc-a0fb-4c97-ae47-18336f6f04cb

It's acting very strange. It's almost as if a couple of joint axes have been made free or something.

It does becomes obvious when throwing around even the non-broken rockets that the placement of the rocket tip body and its joint is not ideal though, as you get a sort of fulcrum-like wobble from having the joint placed so far away from the center-of-mass of the rocket tip. Once you move the joint up to somewhere around where the rocket body meets the rocket tip that this artifact seems to go away. Doing so also happens to resolve the issue with the 9th rocket, but considering how fragile this repro seems to be I would hesitate to draw any conclusions from this.

@jrouwe I would appreciate it if you could take a peek at this, or maybe advise me as to where I can take a closer look, as I'm starting to run out of ideas. I wish I had a more succinct repro for you, but with how sensitive this issue is with regards to all the parameters, timings and initial conditions I frankly don't know how to distill it down to something like a `Samples` test. Seeing as how this is happening even with `FixedConstraint` I'm struggling to think of places where I could have messed this up on my end.

**Reply by Conz3D** [reply]
Nice observation. It make sense to place the joint at the correct position, where the bodies are joined. With that I can't reproduce the bug in my test scene (only a short test sequence). That seems to fix it for me.
But nontheless there is something fishy with the joint on instantiation, IMHO it shouldn't happen in the first place.
Thank you for your tests.

**Reply by Conz3D** [reply]
After some more test, it's not 100% fixed. The bug is less often and if there is a bug, the offset is much smaller. But it still happens.

**Reply by jrouwe** [reply]
I'll take a look at it.

**Reply by mihe** [reply]
@Conz3D You think you could try doing this in the `start()` method of your `rocket.gd`?

```gdscript
rigid_body_3d_tip.top_level = true
```

By pure chance I stumbled upon a bug in Godot where if you have a `RigidBody3D` as a child of another `RigidBody3D` it'll result in a feedback loop that continously sets the physics server transform of the child body every tick, which also wakes it up, preventing the body (and presumably everything it's in contact with) from going to sleep.

This is supposed to be dealt with by the `Node3D::set_ignore_transform_notification` seen [here](https://github.com/godotengine/godot/blob/1ba920fada9efc8c4476ded50fe673b8db541366/scene/3d/physics_body_3d.cpp#L488-L490), but it seems like that ignore only applies to that specific `RigidBody3D`, so when it propagates this transform change to its children then any `RigidBody3D` child will still queue up its transform notification, resulting in the feedback loop.

The only way to get around this is to use `top_level` on the `RigidBody3D` child, which stops this transform propagation from happening.

While this is obviously a problem in itself, I wasn't expecting it to have any effect on the issue at hand, but with that `top_level` thing mentioned above I actually can't reproduce this issue anymore.

I'm still none the wiser as to why it affects the joints in the way that it does, or why it's so random.

**Comment by jrouwe** [comment]
I did some investigations and I think I have a reasonable idea of what's going on. First a quick lesson in physics / fixed constraints:

The physics simulation consists of 2 steps:
* Solve Velocities: This is an iterative solver that solves the relative velocities of the bodies in such a way that the constraint between those 2 bodies is satisfied (e.g. for a fixed constraint it makes the linear and angular velocities the same for both bodies)
* Solve Positions: Because of limited float precision, inevitably the bodies will drift apart even though their velocities are the same. In order to correct this we do another solve step that ensures that the positions are correct again. The big gotcha here is that Jolt (unlike Godot physics) only updates the position of the bodies and does not update the velocities to match the updated positions. The advantage of this is that position correction will not introduce energy in the system and keeps the simulation stable longer. The disadvantage is that if you 'break' a constraint by moving an object like ```pick_anchor.global_position = project_position(event.position, pick_depth)``` (see free_look_camera.gd), Jolt has no idea that this happened. Since the velocity of both objects is unchanged, Solve Velocities will do nothing for 10 iterations. Solve Position only runs 2 iterations and it only solves 20% of the error by default:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/cb2241e1-dadf-4b95-8d1d-ba20a9dc62d9)

This results in the top detaching from the rocket. If I up the position iterations to 10 and set the position correction to 100% (warning this will cause unstability) then you can see that the top follows the rocket much better.

There is another effect at play here and that is that fixed constraints actually consist of 2 constraints: A constraint that welds 2 positions together and a constraint that welds 2 orientations together. These 2 constraints can start working against each other. Constraining the rotation between 2 bodies is simple: just update the angular velocity of both bodies, but constraining the position of the bodies involves updating both the linear and angular velocity of both bodies\*\* and can cause the rotation constraint to break again. Since the solver is iterative, it will solve position, rotation, position, rotation etc. until it hopefully stabilizes after N iterations. In some cases both effects are opposite and cancel so that the system doesn't converge. This especially happens if the pivot point is chosen so that it is far away from the center of mass:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/48b9591f-aa9c-402b-9456-99c9882cd207)

So concluding:
* Move the pivot point up towards the center of mass of the top. If you connect a heavy to a light object, make sure the pivot point is closer to the light object to minimize the error.
* Change the picking code so that it doesn't directly modify the position of a body, make a Kinematic body and move that using velocities to where the mouse is. Use a constraint to constrain the picked body to the Kinematic body.

Note that ```FixedConstraint``` can automatically calculate the pivot point and will use the mass trick, but my guess is that Godot Jolt provides a pivot point so bypasses this code:

https://github.com/jrouwe/JoltPhysics/blob/615444cb3106918d5b756492e36e91d36c1cbd42/Jolt/Physics/Constraints/FixedConstraint.cpp#L76-L91

\*\* Unless the pivot point is on the center of mass for both objects in which case only linear velocity needs to be modified.

**Reply by mihe** [reply]
Thank you for the comprehensive breakdown.

I have a few quick comments, to clear up any potential misunderstandings:

1. I added the picking in my version of the repro. It was not part of the original report or repro. I apologize for adding to the confusion.
2. While the picking may have been implemented incorrectly, I don't think that explains the inconsistent behavior between the different rockets, which was the actual problem, and which my `nine_pickable_rockets.zip` repro did a poor job of showcasing.
3. I believe @Conz3D already [tested](https://github.com/godot-jolt/godot-jolt/discussions/654#discussioncomment-7651988) moving the pivot point closer to the lighter rocket tip body and saw an improvement to the original issue of joints randomly creating a weird "offset", but not a complete fix.
4. Bumping up the solver position iterations to 4+ does indeed seem to alleviate the issue to such a degree that I can't tell if it's there anymore.
5. I am indeed running with `mAutoDetectPoint = false`, but I can't remember exactly why. I may need to take another look at that.

Here's another (hopefully better) repro: [four_pickable_rockets.zip](https://github.com/godot-jolt/godot-jolt/files/13462713/four_pickable_rockets.zip)

I managed to shave down the previous repro to just 4 of these rockets, and was able to "put aside" another one (while not breaking the repro) to use as a reference for what it rightfully should look like, which would be the left-hand one in this case

EDIT: I've now also updated the repro to use a `MoveKinematic` picking anchor instead.

~~Please forgive the continued use of a static pick anchor for now, as I seem to be running up against some issues with my use of `MoveKinematic`~~, but this is what this new repro looks like:

https://github.com/godot-jolt/godot-jolt/assets/4884246/b859851e-f94c-449e-890d-36710eaa0c48

You might be able to see when I'm shaking the right-hand one up-and-down that it creates this same "offset" that @Conz3D was describing in his original issue description. Instead of the rocket tip dragging behind the rocket body, as would be expected due to the points you brought up, the rocket tip seems to instead "lead ahead" with a seemingly amplified version of the rocket body's velocity.

The discrepancy between the two rockets shown there is what's so seemingly fragile, where things like the number of rockets, when/where they launch and if/when any of them get freed will break the repro and make them both behave the same.

And here is what it looks like when uncommenting the `tip.top_level = true` that you'll now find in `rocket.gd`, which I talked about in [the reply above](https://github.com/godot-jolt/godot-jolt/discussions/654#discussioncomment-7663391), that I managed to post just seconds before you posted yours:

https://github.com/godot-jolt/godot-jolt/assets/4884246/534e3a8d-cfb0-4fa8-be77-474ecb76d399

So, I guess this is either...

1. ... still just a manifestation of the points you brought up with regards to the position/velocity solving.
2. ... or maybe some issue caused by the constant calling of `SetPositionAndRotation` and `ActivateBody` for the rocket tip body, caused by the feedback loop mentioned in [my reply above](https://github.com/godot-jolt/godot-jolt/discussions/654#discussioncomment-7663391).
3. ... or this fix for the feedback loop just happens to break the repro and hide the problem, like so many other things
4. ... in which case this might be some other bug maybe?

Let me know you what you think.

**Reply by jrouwe** [reply]
I think it is related to the feedback loop, it causes a tiny error that apparently gets amplified (I don't know yet how).

I added the following debug code:

```
void BodyInterface::SetPositionAndRotation(const BodyID &inBodyID, RVec3Arg inPosition, QuatArg inRotation, EActivation inActivationMode)
{
	static atomic<int> accept { 0 };
	if (++accept > 2000)
	{
		std::cout << "No longer accepting position updates" << std::endl;
		return;
	}
...
```

If you do that and start shaking the right rocket, you will see that as soon as it starts tracing 'No longer accepting position updates' the right rocket starts behaving as the left rocket.

Furthermore I added the following code:

```
void Body::SetPositionAndRotationInternal(RVec3Arg inPosition, QuatArg inRotation, bool inResetSleepTestSpheres)
{
	JPH_ASSERT(BodyAccess::sCheckRights(BodyAccess::sPositionAccess, BodyAccess::EAccess::ReadWrite));

	Vec3 oldpos = mPosition;
	Quat oldrot = mRotation;

	mPosition = inPosition + inRotation * mShape->GetCenterOfMass();
	mRotation = inRotation;

	Vec3 delta = mPosition - oldpos;
	if (delta.Length() > 0.001f)
		std::cout << "Moved body from " << ConvertToString(oldpos).c_str() << " to " << ConvertToString(mPosition).c_str() << " by " << delta << std::endl;
	Quat deltar = (mRotation - oldrot);
	if (deltar.Length() > 0.001f)
		std::cout << "Rotated body from " << ConvertToString(oldrot).c_str() << " to " << ConvertToString(mRotation).c_str() << " by " << deltar << std::endl;
...
```

If you start shaking the left rocket, you'll get 1 trace for both corresponding to the creation of the body used for picking. After that it is silent.

If you touch the 2nd rocket, you'll get a constant stream of traces that look like this:

```
Moved body from -4.78575, 0.684944, -1.57593 to -4.79354, 0.685154, -1.57844 by -0.00779533, 0.000209987, -0.00250208
Rotated body from 0.722366, -0.545393, -0.21347, -0.367647 to 0.72284, -0.546271, -0.211634, -0.366472 by 0.000473261, -0.000878513, 0.00183606, 0.00117484
```

I can see that it comes from:

```
>	godot-jolt_windows-x64_editor.dll!JPH::Body::SetPositionAndRotationInternal(JPH::Vec3 inPosition, JPH::Quat inRotation, bool inResetSleepTestSpheres) Line 115	C++
 	godot-jolt_windows-x64_editor.dll!JPH::BodyInterface::SetPositionAndRotation(const JPH::BodyID & inBodyID, JPH::Vec3 inPosition, JPH::Quat inRotation, JPH::EActivation) Line 394	C++
 	godot-jolt_windows-x64_editor.dll!JoltObjectImpl3D::_apply_transform(const godot::Transform3D & p_transform, bool p_lock) Line 454	C++
 	godot-jolt_windows-x64_editor.dll!JoltObjectImpl3D::set_transform(godot::Transform3D p_transform, bool) Line 120	C++
 	godot-jolt_windows-x64_editor.dll!JoltBodyImpl3D::set_state(godot::PhysicsServer3D::BodyState p_state, const godot::Variant & p_value) Line 94	C++
 	godot-jolt_windows-x64_editor.dll!godot::call_with_ptr_args_helper<JoltPhysicsServer3D,godot::RID const &,enum godot::PhysicsServer3D::BodyState,godot::Variant const &,0,1,2>(JoltPhysicsServer3D * p_instance, void(JoltPhysicsServer3D::*)(const godot::RID &, godot::PhysicsServer3D::BodyState, const godot::Variant &) p_method, const void * const * p_args, IndexSequence<0,1,2> __formal) Line 194	C++
 	[External Code]	
```

So I think it's from the feedback loop. The delta position is in the order of millimeters, so that would definitively be enough to violate the constraint and for Jolt to have to fall back on the Position Solve step. 

So it looks like there is some sort of error in the feedback loop that results in a delta position / rotation. Why it's only the right rocket and not the left I don't know (my godot knowledge stops here).

**Reply by mihe** [reply]
> I added the following debug code:
> 
> [...]
> 
> If you do that and start shaking the right rocket, you will see that as soon as it starts tracing 'No longer accepting position updates' the right rocket starts behaving as the left rocket.

Nice catch! I would agree that this more or less confirms it to have been the issue.

> Why it's only the right rocket and not the left I don't know (my godot knowledge stops here).

I would be tempted to do some more digging to see why that is exactly, but I think I'll just accept this at face value and leave that as an unsolved mystery.

I'll see if it's possible to create a PR to fix this feedback loop issue in Godot, but until then @Conz3D you'll just have to rely on the `top_level` property for any `RigidBody3D` that's a child (direct or indirect) of another `RigidBody3D`.

Unfortunately `top_level` also impacts its transform in the editor, where it won't follow if you move the parent body, making it somewhat annoying to work with if you just enable it in the property inspector in Godot, so I would recommend you enable it in your `start()` method instead, as mentioned in [my reply above](https://github.com/godot-jolt/godot-jolt/discussions/654#discussioncomment-7663391).

**Reply by Conz3D** [reply]
Thank you both for your bug hunt. I think I have to read your answers more than once, to grasp them fully.
The solution with "top_level" will only work in this spec. case, right? If I have a little bit more complex system, that solution wouldn't work, I think.
Like three rigid bodies connected by two joints:
. RB (RigidBody)
.. J (Joint)
... RB
....J
.....RB
Only one of the connected RigidBodies could be the "top_level", am I right?

**Reply by mihe** [reply]
> Only one of the connected RigidBodies could be the "top_level", am I right?

No, I believe you should be able to use `top_level` on all of them.

I would maybe even go so far as to argue that _every_ `RigidBody3D` should be made top-level in their `_ready` method, to prevent any weird/unintuitive transform propagation from their parent node. The only potential downside to this that I can see is that you won't be able to conveniently get its local-space transform if needed, but you still have access to the parent, so that should still be doable through something like `get_parent().global_transform.affine_inverse() * global_transform`.

The joints' placements in the scene tree hierarchy have no bearing on any of this, as far as I know, and shouldn't need to be `top_level`.

**Comment by Conz3D** [comment]
The solution with `top_level` on `_ready` is working good so far. Not one tip with offset in my tests. 
One little glitch. Now I can see the tip for maybe one frame at the base of the rocket (lower end of the stick) on instantiation. It jumps to the top in the next frame. But than it is connected there as expected without the bug.

**Reply by mihe** [reply]
> Now I can see the tip for maybe one frame at the base of the rocket (lower end of the stick) on instantiation

Does it make any difference if you set `top_level` in `_init` instead?

**Comment by Conz3D** [comment]
> > Now I can see the tip for maybe one frame at the base of the rocket (lower end of the stick) on instantiation
> 
> Does it make any difference if you set `top_level` in `_init` instead?

With `top_level` in `_init`, I get following error:
`Invalid set index 'top_level' (on base: 'Nil') with value of type 'bool'.`

**Reply by mihe** [reply]
Oh right, you're doing this from the parent. You need to do this from the actual node itself, so a separate script for the rocket tip's `RigidBody3D`.

**Reply by Conz3D** [reply]
ok, with that I don't get the error msg on `_init`. But the tip is still visible for 1-2 frames at the bottom. So no fix for the glitch.

**Reply by mihe** [reply]
What if you also in your `RocketLaunchPad` (or whatever equivalent you have in your actual application) set the position/rotation of the rocket _before_ you add the rocket as a child?

**Reply by Conz3D** [reply]
YAY! That fixed it. Thank you so much for your help. On my end it seems to be working now without bug or glitch.
With these insights, there are some settings you have to follow to get a working setup with joints. IMHO, it should noted somewhere in the docs. Do you plan to fix this in code on your side somehow, to make it less errorprone for the users?
EDIT:
I'll try to test it in `_ready` and on the parent again, now that one problem was to set position after adding as child, to see if these are necessary settings.
EDIT2:
Yep, it's working on the parent in `_ready`, too. So it's not a necessery setup to have it on the RB itself or on `_init`.

**Reply by mihe** [reply]
> YAY! That fixed it. Thank you so much for your help. On my end it seems to be working now without bug or glitch.

Finally. :)

> IMHO, it should noted somewhere in the docs. Do you plan to fix this in code on your side somehow, to make it less errorprone for the users?

None of this has anything to do with this extension, technically speaking, and happens even if you use Godot Physics and the default joints. Although, the issue with `RigidBody3D` being the child of another `RigidBody3D` doesn't manifest in quite the same way in Godot Physics. Either way, it's a bug in Godot itself, which I have a fix for lined up already, but I'll wait with making a PR until after 4.2-stable is out, since the people who would need to review it are likely quite busy with the upcoming release right now.

Another potential workaround to that problem might be to just avoid having any `RigidBody3D` being the child of another `RigidBody3D`. So in your case it might work to have it look something like:

```
- Node3D (Rocket)
  - RigidBody3D (Rocket Body)
    - Shapes
    - Sounds
    - Particles
  - RigidBody3D (Rocket Tip Body)
    - Shapes
  - Joint (This can be wherever)
```

But that comes with the drawback that your root `Node3D` will never move, so querying for the transform of your instantiated scene is going to be weird.

With regards to the issue of needing to set position/rotation before `add_child`, that's something you'll see with Godot Physics as well. I'm not entirely sure what's causing this, but presumably it's got something to do with the fact that setting position/rotation after `add_child` has been called means you're also moving the rocket body _after_ the rocket tip's `top_level` has come into effect, which obviously means the tip shouldn't move.

What I don't quite understand is why the tip glitches back to its intended position. I would've thought it would just stay at its initial position, but maybe the joint manages to inadvertently correct it somehow.

> Yep, it's working on the parent in `_ready`, too. So it's not a necessery setup to have it on the RB itself or on `_init`.

Cool, good to know.
